### PR TITLE
Fishy swarmers: Carp content and Space Dragon!

### PR DESCRIPTION
**Pull request by CrudeWax**
## About the PR 
A PR expanding on the space carp fauna, adding a multitude of content related to beloved space fish.

**Space Dragon!**
The main big bad. A huge mob which has a 500 HP heatlhpool, and a 45 slow melee hit which can't arc.
The dragon is able to devour human corpses to heal and gain charges on his carp-spawning ability, creating sentient carps indistinguishable from regular ones beyond their ability to speak.
The dragon possesses very unqie biological properties- it's metabolism is 4 times faster than that of a human, making him very vulnerable to any chemicals piercing it's hide and entering it's bloodstream.

**New chemicals!**
Nerfs heartstopper toxin, making it deal only 2 damage of asphyxiation per tick, effectively halving it's effectiveness.

Adds **Ichor**, the rare chemical that dragon gets when it devours a corpse, extremely potent healing agent, it can very easily knit bleeding wounds, rapidly healing brute and burn damage. The humans also get a chance to use it, because after dragon's death, it drops meat rich with ichor contents.

Adds **CarpoToxin** or CarpToxin. A small amount of carp toxin is stored in carp fillets, which you get 2 pieces of when butchering a space carp. Carpotoxin is a potent poison, dealing the heartstopper's old 4 damage per tick, only now with poison, and while alerting the target of their poisoning via intense burning sensations. 

**Exotic Dishes!**

The only way to describe the strong taste of CarpToxin is "Exotic", that includes sashimi and cuban carp dishes, which are now cook-able and filled with lethal doses of CarpToxin, however there is an unlikely antitode- Carpentium. Because they both have "Carp" in the name, obviously.

**Screenshots**

image
image
image
image
image
image

**VIDEO:**
https://youtu.be/8Rflxm5nDRQ

**Changelog**

:cl: CrudeWax
- add: Nanotrasen's deeper research into the space-faring carpfish has proven that their meat contains extremely potent poison!
- add: New carp meat dishes are on the menu for experienced cooks! Their taste can only be described as "Exotic", by those willing to stomach their lethal effects.
- add: It seems like the carpfish kind also have an apex predator among their ranks, be warned of large influxes of organic activity and leviathan space-faring creatures
- add: Some carp appear to be smarter than others and capable of humanlike speech
- tweak: due to overuse of heartstopper toxin, and it's more daring carp alternative, the effects of the former have been greatly reduced
- fix: Space-faring carp will no longer die of being exposed to the coldness of space

### CODE REVIEW

File: Content.Server/Dragon/DragonComponent.cs
Comment (DogZeroX) [inline Content.Server/Dragon/DragonComponent.cs:12]:
```suggestion
        [DataField("voreAction", required: true)]
```

**Comment by CrudeWax** [comment]
carpmeatMeme

**Comment by CrudeWax** [comment]
The bulk of the content is complete, and the only things left to finish are minor additions like sounds, icons and ideaguy-ing stuff.

**Comment by CrudeWax** [comment]
Main question for maintainers and admins:

SHOULD THERE BE AN ORGANIC INFLUX EVENT? (an event to randomly spawn dragon) and how rare should it be?

**Comment by CrudeWax** [comment]
I have decided against adding organic influx as to not surprise admins with a 500 HP behemoth.

File: Content.Server/Dragon/DragonComponent.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonComponent.cs:10]: for the future we use filescoped namespaces now so you dont need to have a namespace block

File: Content.Server/Dragon/DragonSystem.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonSystem.cs:44]: is this a todo for a future PR or here? this shouldn't be too hard to do; butchering gibs the entity, and theres an event for before gibbing happens where you can handle this behavior.

File: Content.Server/Dragon/DragonSystem.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonSystem.cs:63]:
```suggestion
        private void OnDevourSuccessful(TargetDevourSuccessfulEvent args)
```

and event handler change

File: Content.Server/Dragon/DragonSystem.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonSystem.cs:137]: WAY too indented you need to invert the control flow here in some places (`if (!comp) return;` rather than `(if comp) { } else { return; }`)

File: Content.Server/Dragon/DragonSystem.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonSystem.cs:68]: this needs to be made more obvious that this is only for non-mobs, got kind of confused since the two different kind of devours arent differentiated much by name

File: Content.Server/Dragon/DragonComponent.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonComponent.cs:71]: reduce whitespace here

File: Content.Server/Dragon/DragonSystem.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonSystem.cs:142]: use `TagSystem` methods rather than directly accessing the tag hashset

Comment (mirrorcult) [inline Content.Server/Dragon/DragonSystem.cs:142]: this is a weird way to tag it but why not. although it should really be using an `EntityWhitelist` on the component, rather than hardcoding the tag here, and doing `IsValid()` rather than `HasTag` or `Tags.Contains`. that way this can be tweaked in the future if we need to very easily

File: Content.Server/Dragon/DragonSystem.cs
Comment (mirrorcult) [inline Content.Server/Dragon/DragonSystem.cs:166]:
```suggestion
                Spawn(component.CarpProto, Transform(dragonuid).Coordinates);
```

File: Resources/Locale/en-US/actions/actions/dragon.ftl
Comment (mirrorcult) [inline Resources/Locale/en-US/actions/actions/dragon.ftl:2]:
```suggestion
devour-action-popup-message-fail-target-not-valid = That doesn't look particularly edible.
```

File: Resources/Locale/en-US/actions/actions/dragon.ftl
Comment (mirrorcult) [inline Resources/Locale/en-US/actions/actions/dragon.ftl:1]:
```suggestion
devour-action-popup-message-structure = Your jaws dig into thick material..
```

File: Resources/Locale/en-US/actions/actions/dragon.ftl
Comment (mirrorcult) [inline Resources/Locale/en-US/actions/actions/dragon.ftl:3]:
```suggestion
devour-action-popup-message-fail-target-no-body = Your palate isn't ready for whatever that is.
```

File: Resources/Locale/en-US/actions/actions/dragon.ftl
Comment (mirrorcult) [inline Resources/Locale/en-US/actions/actions/dragon.ftl:4]:
```suggestion
devour-action-popup-message-fail-target-alive = You can't consume creatures that are alive!
```

File: Resources/Locale/en-US/actions/actions/dragon.ftl
Comment (mirrorcult) [inline Resources/Locale/en-US/actions/actions/dragon.ftl:12]:
```suggestion
action-name-carp-birth = Summon carp
```

File: Resources/Prototypes/Entities/Mobs/NPCs/carp.yml
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Mobs/NPCs/carp.yml:46]: leave the comment its funny!!!!!

Comment (mirrorcult) [reply Resources/Prototypes/Entities/Mobs/NPCs/carp.yml:46]: and soul etc etc

File: Resources/Prototypes/Entities/Mobs/Player/dragon.yml
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Mobs/Player/dragon.yml:22]: does it really need that

File: Resources/Prototypes/Entities/Mobs/Player/dragon.yml
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Mobs/Player/dragon.yml:120]: this should have a higher delay, maybe 30secs

File: Resources/Prototypes/Entities/Mobs/Player/dragon.yml
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Mobs/Player/dragon.yml:114]: theres alread ya whitelist here like i mentioned before that you needed, so just make it `MobState` and `RCDDeconstructWhitelist` or whatever you were using before

File: Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml:15]: dont add that to base wall

File: Resources/Prototypes/Entities/Structures/Doors/Airlocks/base_structureairlocks.yml
Comment (mirrorcult) [inline Resources/Prototypes/Entities/Structures/Doors/Airlocks/base_structureairlocks.yml:53]: im sus on adding it to base airlock, this may be getting overriden and not properly tagged by its children as well, since the tag dicts wont merge

File: Resources/Prototypes/Reagents/biological.yml
Comment (mirrorcult) [inline Resources/Prototypes/Reagents/biological.yml:54]: i was gonna say add a condition for dragon metabolizertype BUT its probably more fun if anyone can eat this stuff since its so rare

File: Resources/Prototypes/Reagents/medicine.yml
Comment (mirrorcult) [inline Resources/Prototypes/Reagents/medicine.yml:486]: these are already getting healed by the groups, no need to add them

File: Resources/Prototypes/Reagents/toxins.yml
Comment (mirrorcult) [inline Resources/Prototypes/Reagents/toxins.yml:22]:
```suggestion
  id: Carpotoxin
```

consistency

**Review by mirrorcult — CHANGES_REQUESTED** [review]
can you add a video of it working?

File: Resources/Prototypes/Reagents/biological.yml
Comment (CrudeWax) [reply Resources/Prototypes/Reagents/biological.yml:54]: It comes in the dragon's meat as a reward for killing him

File: Resources/Prototypes/Entities/Mobs/Player/dragon.yml
Comment (CrudeWax) [reply Resources/Prototypes/Entities/Mobs/Player/dragon.yml:120]: It's limited by summon tokens so you won't have an ability to spam it constantly.

File: Resources/Prototypes/Entities/Mobs/Player/dragon.yml
Comment (CrudeWax) [reply Resources/Prototypes/Entities/Mobs/Player/dragon.yml:22]: Otherwise it gets the disarm action

File: Resources/Prototypes/Entities/Structures/Doors/Airlocks/base_structureairlocks.yml
Comment (CrudeWax) [reply Resources/Prototypes/Entities/Structures/Doors/Airlocks/base_structureairlocks.yml:53]: How else to adress it, given airlocks do need to be deconstructed by an RCD? (Unless we don't want that)

File: Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml
Comment (CrudeWax) [reply Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml:15]: Why and how else?

**Comment by CrudeWax** [comment]
Video here
https://youtu.be/8Rflxm5nDRQ

File: Content.Server/Dragon/DragonSystem.cs
Comment (metalgearsloth) [reply Content.Server/Dragon/DragonSystem.cs:137]: Also please don't inline else statements.

File: Content.Server/Dragon/DragonSystem.cs
Comment (metalgearsloth) [reply Content.Server/Dragon/DragonSystem.cs:68]: Also pass in entitymanager to the filter + use the proxy method for queuedeleteentity (queuedel iirc).

File: Resources/Prototypes/Entities/Mobs/Player/dragon.yml
Comment (metalgearsloth) [reply Resources/Prototypes/Entities/Mobs/Player/dragon.yml:22]: Adding a dummy action isn't a good solution, make it null or something.

File: Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml
Comment (metalgearsloth) [reply Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml:15]: After prototype composition is merged via that, or just add it to walls as needed.

File: Resources/Prototypes/Entities/Mobs/NPCs/carp.yml
Comment (CrudeWax) [reply Resources/Prototypes/Entities/Mobs/NPCs/carp.yml:46]: My friendship with crapmeat is over.
Meatfish is my new best friend.

File: Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml
Comment (mirrorcult) [reply Resources/Prototypes/Entities/Structures/Walls/base_structurewalls.yml:15]: add it to the children walls

File: Content.Server/Dragon/DragonSystem.cs
Comment (CrudeWax) [reply Content.Server/Dragon/DragonSystem.cs:68]: I renamed the method but what proxy method and what soundsystem const?

**Comment by CrudeWax** [comment]
Still being worked on.

File: Resources/Prototypes/Entities/Mobs/Player/dragon.yml
Comment (CrudeWax) [reply Resources/Prototypes/Entities/Mobs/Player/dragon.yml:114]: Does not work! I don't know why, but for whatever reason unless I use "Tag" it won't consider walls as vallid

**Comment by metalgearsloth** [comment]
I cleaned up a bunch of code but I'ma be real, the dragon sprite looks bad.

**Comment by EmoGarbage404** [comment]
> I cleaned up a bunch of code but I'ma be real, the dragon sprite looks bad.

It's not the worst thing imaginable, but it's definitely feeling that 64x64 sprite restriction. It's got some SOVL tbh.

**Comment by metalgearsloth** [comment]
> > I cleaned up a bunch of code but I'ma be real, the dragon sprite looks bad.
> 
> It's not the worst thing imaginable, but it's definitely feeling that 64x64 sprite restriction. It's got some SOVL tbh.

Mirino showed me the other ones on /tg/ and most of them look better.

**Comment by CrudeWax** [comment]
Firstly this sprite is taken off /TG/
second, what's the issue?

**Comment by CrudeWax** [comment]
Ready for merge

**Comment by metalgearsloth** [comment]
Needs audio crediting

**Comment by metalgearsloth** [comment]
@CrudeWax You coming back to this? Only thing I can't fix is the audio licensing.

**Pull request merged**
